cocotb/types/range.py

```diff
@@ -0,0 +1,178 @@
+            right, direction = direction, None
+        if direction is None:
+            step = 1 if left < right else -1
+        else:
+            step = self._translate_direction(direction)
+        self._range = range(left, right + step, step)
+
+    @staticmethod
+    @cache
+    def _translate_direction(direction) -> int:
+        direction = direction.lower()
+        if direction == "to":
+            return 1
+        elif direction == "downto":
+            return -1
+        else:
+            raise ValueError("direction must be 'to' or 'downto'")
+
+    @classmethod
+    def from_range(cls, rng: range) -> "Range":
+        """Converts :class:`range` to :class:`Range`."""
+        if rng.step not in (1, -1):
+            raise ValueError("step must be 1 or -1")
+        obj = cls.__new__(cls)
+        obj._range = rng
+        return obj
+
+    def to_range(self) -> range:
+        """Convert :class:`Range` to :class:`range`."""
+        return self._range
+
+    @cached_property
+    def left(self) -> int:
+        """Leftmost value in a range."""
+        return self._range.start
+
+    @cached_property
+    def direction(self) -> str:
+        """``'to'`` if values are meant to be ascending, ``'downto'`` otherwise."""
+        return "to" if self._range.step == 1 else "downto"
+
+    @cached_property
+    def right(self) -> int:
+        """Rightmost value in a range."""
+        return self._range.stop - self._range.step
+
+    @cached_property
+    def length(self) -> int:
+        """Length of range."""
+        return len(self._range)
+
+    def __len__(self) -> int:
+        return self.length
+
+    @overload
+    def __getitem__(self, item: int) -> int:
+        pass  # pragma: no cover
+
+    @overload
+    def __getitem__(self, item: slice) -> "Range":
+        pass  # pragma: no cover
+
+    def __getitem__(self, item):
+        if isinstance(item, int):
+            return self._range[item]
+        elif isinstance(item, slice):
+            return type(self).from_range(self._range[item])
+        raise TypeError(
+            "indices must be integers or slices, not {}".format(type(item).__name__)
+        )
+
+    def __contains__(self, item: Any) -> bool:
+        return item in self._range
+
+    def __iter__(self) -> Iterator[int]:
+        return iter(self._range)
+
+    def __reversed__(self) -> Iterator[int]:
+        return reversed(self._range)
+
+    def __eq__(self, other: Any) -> bool:
+        if type(self) is not type(other):
+            return NotImplemented
+        return self._range == other._range
+
+    def __hash__(self) -> int:
+        return hash(self._range)
+
+    def index(self, item: Any) -> int:
+        return self._range.index(item)
+
+    def count(self, item: Any) -> int:
+        return self._range.count(item)
+
+    def __repr__(self) -> str:
+        return "{}({!r}, {!r}, {!r})".format(
+            type(self).__name__, self.left, self.direction, self.right
```

Comment:
```suggestion
            type(self).__qualname__, self.left, self.direction, self.right
```

Reply:
Looks like this is preferable. https://github.com/ericvsmith/dataclasses/issues/100

Reply:
Looks like this is preferable https://github.com/ericvsmith/dataclasses/issues/100.